FWI_forward_adjoint/FWIpart1.md

```diff
@@ -168,53 +171,68 @@ and we see that at time step `i` we need to add the source term corresponding to
 
 ```python
-	# create receiver array from receiver coordinates
-	rec = Receiver(name='rec', npoint=101, ntime=nt, ndim=2, coordinates=rec_coords)
-	rec_term = rec.interpolate(u, offset=model.nbpml)
+  # Create receiver array from receiver coordinates
+  rec = Receiver(name='rec', grid=model.grid, npoint=101, ntime=nt, coordinates=rec_coords)
+  rec_term = rec.interpolate(u, offset=model.nbpml)
 ```
 
 and the `offset` parameter also correct for the origin shift from the model extension.
 
 ### Forward simulation 
 
-With the source/receiver geometry set and the wave-equation stencil generated, we can now define our forward propagator by adding the source and receiver terms to our stencil object; that is, in `Python` we have:
+With the source/receiver geometry set and the wave-equation stencil
+generated, we can now define our forward propagator by adding the
+source and receiver terms to our previously defined `stencil`
+object. The resulting list is used to generate a `Operator` object
+that performs automated generation and optimization of low-level C
+code according to the symbolic expressions provided:
 
 ```python
-	# Create forward propagator
-	op_fwd = Operator([stencil] + src_term + rec_term,
-	              subs={t.spacing: dt, x.spacing: spacing[0],
-	                    y.spacing: spacing[1]})
+  # Create forward propagator
+  op_fwd = Operator([stencil] + src_term + rec_term)
 ```
 
-The Devito operator creation is minimalist thanks to the symbolic representation of the stencil. All information such as dimension sizes (to generate loops), are contained in the stencil and more specifically in its arguments `u, m, dam, src, rec` that still carry the metadata provided at object instantiation (`TimeData, DenseData, ...` creation). The only extra argument is `subs` that provides substitution method for the constants. At generation time, all instances of `x.spacing=h_x` will be replaced by its actual value. Once the operator created, we can access the generated C code with `op_fw.ccode`.
-
-####Figure:{#Cgen}
-![Generated C code](Figures/ccode-crop.png)
-
-We can finally execute the forward modeling propagator with the simple command:
+The symbolic expressions used to create `Operator` contain sufficient
+meta-information for Devito to create a fully functional computational
+kernel. The dimension symbols contained in the symbolic function
+object (`t, x, y, z`) define the loop structure of the created code,
+while allowing Devito to automatically optimize the underlying loop
+structure to increase performance.
+
+The size of the loops and spacing between grid points is inferred from
+the symbolic `Function` objects and associated `model.grid` object at
+run-time. As a result we can invoke the generated kernel through a
+simple Python function call by supplying the number of timesteps
+`time` and the timestep size `dt`. The user data associated with each
```

Comment:
This is also why I didn't go with the new API. This kind of details to be required as arguments is not very "user-friendly" in my opinion. I still prefer substitutions at generation.

Reply:
Yeah well, what can I say, the old ways still work. IIRC, the substitutions were flagged by @flexie in the call, so take it or leave it.

Reply:
Yes he did, but more for explanatory reason. I guaranty you that if we start telling him we don't hard code constants anymore, the conf call is gonna be a lot more than 2h and a lot less pleasant, not matter the reason we don't.